File: frontend/src/api/authorBible.ts
// AuthorBible API 客户端：读取和保存项目级正式写作规则、风格指南、约束和初始世界状态。
import { getData, putData } from './http';
import type { AuthorBible, UpdateAuthorBibleRequest, WorldStateEntry } from '../types/api';

type WorldStateEntryPayload = WorldStateEntry & {
  Key?: string;
  Value?: unknown;
  Note?: string;
};

type AuthorBiblePayload = AuthorBible & {
  ID?: string;
  ProjectID?: string;
  Theme?: string;
  StyleGuide?: string;
  WorldRules?: string[];
  AestheticPrinciples?: string[];
  HardConstraints?: string[];
  SoftPreferences?: string[];
  ForbiddenMoves?: string[];
  InitialWorldState?: WorldStateEntryPayload[];
};

export async function getAuthorBible(projectId: string, signal?: AbortSignal) {
  return normalizeAuthorBible(await getData<AuthorBiblePayload>(`/projects/${projectId}/author-bible`, signal));
}

export async function updateAuthorBible(projectId: string, body: UpdateAuthorBibleRequest) {
  return normalizeAuthorBible(
    await putData<AuthorBiblePayload, UpdateAuthorBibleRequest>(`/projects/${projectId}/author-bible`, body),
  );
}

function normalizeAuthorBible(bible: AuthorBiblePayload): AuthorBible {
  const worldState = bible.initial_world_state ?? bible.InitialWorldState ?? [];

  return {
    ...bible,
    id: bible.id ?? bible.ID,
    project_id: bible.project_id ?? bible.ProjectID,
    theme: bible.theme ?? bible.Theme,
    style_guide: bible.style_guide ?? bible.StyleGuide,
    world_rules: bible.world_rules ?? bible.WorldRules ?? [],
    aesthetic_principles: bible.aesthetic_principles ?? bible.AestheticPrinciples ?? [],
    hard_constraints: bible.hard_constraints ?? bible.HardConstraints ?? [],
    soft_preferences: bible.soft_preferences ?? bible.SoftPreferences ?? [],
    forbidden_moves: bible.forbidden_moves ?? bible.ForbiddenMoves ?? [],
    initial_world_state: worldState.map(normalizeWorldStateEntry),
  };
}

function normalizeWorldStateEntry(entry: WorldStateEntryPayload): WorldStateEntry {
  return {
    key: entry.key ?? entry.Key ?? '',
    value: entry.value ?? entry.Value,
    note: entry.note ?? entry.Note,
  };
}

File: internal/transport/http/handler/author_bible.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/fishimei/NovelOS/internal/application/model"
	"github.com/fishimei/NovelOS/internal/application/port"
	"github.com/fishimei/NovelOS/internal/transport/http/dto"
	"github.com/fishimei/NovelOS/internal/transport/http/presenter"
)

type AuthorBibleHandler struct {
	authorBibles port.AuthorBibleRepository
}

func NewAuthorBibleHandler(authorBibles port.AuthorBibleRepository) AuthorBibleHandler {
	return AuthorBibleHandler{authorBibles: authorBibles}
}

func (h AuthorBibleHandler) Get(c *gin.Context) {
	result, err := h.authorBibles.GetByProjectID(c.Request.Context(), c.Param("project_id"))
	if err != nil {
		presenter.Error(c, err)
		return
	}

	presenter.Data(c, http.StatusOK, result)
}

func (h AuthorBibleHandler) Update(c *gin.Context) {
	var req dto.UpdateAuthorBibleRequest
	if !bindJSON(c, &req) {
		return
	}

	worldState := make([]model.WorldStateEntry, 0, len(req.InitialWorldState))
	for _, entry := range req.InitialWorldState {
		worldState = append(worldState, model.WorldStateEntry{
			Key:   entry.Key,
			Value: entry.Value,
			Note:  entry.Note,
		})
	}

	result, err := h.authorBibles.UpdateByProjectID(c.Request.Context(), c.Param("project_id"), model.UpdateAuthorBibleInput{
		Theme:               req.Theme,
		StyleGuide:          req.StyleGuide,
		WorldRules:          req.WorldRules,
		AestheticPrinciples: req.AestheticPrinciples,
		HardConstraints:     req.HardConstraints,
		SoftPreferences:     req.SoftPreferences,
		ForbiddenMoves:      req.ForbiddenMoves,
		InitialWorldState:   worldState,
	})
	if err != nil {
		presenter.Error(c, err)
		return
	}

	presenter.Data(c, http.StatusOK, result)
}
